### PR TITLE
Deprecate type='tuples for chatbot and focus chatbot docs on 'messages' type

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Closes: #9007

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: js/_website/src/lib/templates/gradio/03_components/chatbot.svx [hunk 1]
```diff
@@ -38,21 +38,42 @@ demo.launch()
...
 
+### Message format
 
 If the `type` is `'messages'`, then the data sent to/from the chatbot will be a list of dictionaries
 with `role` and `content` keys. This format is compliant with the format expected by most LLM APIs (HuggingChat, OpenAI, Claude).
-The `role` key is either `'user'` or `'`assistant'` and the `content` key can be a string (markdown/html supported),
-a `FileDataDict` (to represent a file that is displayed in the chatbot - documented below), or a gradio component.
+The `role` key is either `'user'` or `'`assistant'` and the `content` key can be one of the following:
 
+1. A string (markdown/html is also supported).
```
Comment (freddyaboulton) [inline js/_website/src/lib/templates/gradio/03_components/chatbot.svx:50]: @aliabd can you please help me with the formatting of this list?

image

Comment (abidlabs) [reply js/_website/src/lib/templates/gradio/03_components/chatbot.svx:50]: Yeah all of the bullet points seem to be missing. e.g. in the guide:

image

**Comment by abidlabs** [comment]
Looks great @freddyaboulton! Some feedback on the docs:

(1) I don't even think we should mention the lists of lists format here: 

image

I would say something like this:

* If you set `type="messages"` in `gr.ChatInterface`, the history will be a list of dictionaries with role and content keys. Please see the chatbot [docs](https://gradio-jxd4bjx6h-hugging-face.vercel.app/docs/gradio/chatbot) for an in-depth explanation of the chatbot format. 
* Add a note saying that we *always* recommend setting `type="messages"` since `type="tuples"` is deprecated and will be removed in a future version of Gradio
* Also I'd add an example of what `history` looks like with `type="messages"` since if someone is not familiar already, its not totally clear what it looks like


Nit: unrelated to this PR, but can we disable autofocus when 

image

**Comment by abidlabs** [comment]
(3) In a couple of places, we set `type="messages"` in both `gr.Chatbot` and `gr.ChatInterface` -- that should not be required, right?

image

File: gradio/chat_interface.py [hunk 2]
```diff
@@ -92,6 +92,7 @@ def __init__(
...
+            type: The format of the messages. If 'messages', passes the value as a list of dictionaries with 'role' and 'content' keys. The `content` key's value should be one of the following - (1) strings in valid Markdown, (2) tuples if sending files: (a filepath or URL to a file, [optional string alt text]) -- if the file is image/video/audio, it is displayed in the Chatbot, (3) a dictionary with a 'path' key corresponding to the file to display or (4) an instance of a Gradio component. At the moment Image, Plot, Video, Gallery, Audio, and HTML are supported. The 'role' key should be one of 'user' or 'assistant'. Any other roles will not be displayed in the output. If 'tuples', expects a `list[list[str | None | tuple]]`, i.e. a list of lists. The inner list should have 2 elements: the user message and the response message. Each message can be a valid type of the `content` key in the 'messages' format. Additionally, if the message is None, it will not be displayed.
```
Comment (abidlabs) [inline gradio/chat_interface.py:95]: Do we support (2) in the new format? I think we should remove it as its not necessary anymore 
```suggestion
            type: The format of the messages passed into the chat history parameter of `fn`. If "messages", passes the value as a list of dictionaries with openai-style "role" and "content" keys. The "content" key's value should be one of the following - (1) strings in valid Markdown (2) a dictionary with a "path" key and value corresponding to the file to display or (3) an instance of a Gradio component. At the moment Image, Plot, Video, Gallery, Audio, and HTML are supported. The "role" key should be one of 'user' or 'assistant'. Any other roles will not be displayed in the output. If this parameter is 'tuples', expects a `list[list[str | None | tuple]]`, i.e. a list of lists. The inner list should have 2 elements: the user message and the response message, but this format is deprecated.
```

File: gradio/components/chatbot.py [hunk 3]
```diff
@@ -172,7 +173,7 @@ def __init__(
...
+            type: The format of the messages. If 'messages', passes the value as a list of dictionaries with 'role' and 'content' keys. The `content` key's value should be one of the following - (1) strings in valid Markdown, (2) tuples if sending files: (a filepath or URL to a file, [optional string alt text]) -- if the file is image/video/audio, it is displayed in the Chatbot, (3) a dictionary with a 'path' key corresponding to the file to display or (4) an instance of a Gradio component. At the moment Image, Plot, Video, Gallery, Audio, and HTML are supported. The 'role' key should be one of 'user' or 'assistant'. Any other roles will not be displayed in the output. If 'tuples', expects a `list[list[str | None | tuple]]`, i.e. a list of lists. The inner list should have 2 elements: the user message and the response message. Each message can be a valid type of the `content` key in the 'messages' format. Additionally, if the message is None, it will not be displayed.
```
Comment (abidlabs) [inline gradio/components/chatbot.py:176]: Similar suggestions to above

**Comment by abidlabs** [comment]
(4) Can we update these examples to explicitly set `type="messages"`?

image

We might even want to set `type="messages"` explicitly here:

image

File: js/_website/src/lib/templates/gradio/03_components/chatbot.svx [hunk 4]
```diff
@@ -38,21 +38,42 @@ demo.launch()
...
+This parameter can take two values, `'tuples'` and `'messages'`. 
+The `'tuples'` type is deprecated and will be removed in version 6.0 of Gradio.
 
+### Message format
 
 If the `type` is `'messages'`, then the data sent to/from the chatbot will be a list of dictionaries
 with `role` and `content` keys. This format is compliant with the format expected by most LLM APIs (HuggingChat, OpenAI, Claude).
-The `role` key is either `'user'` or `'`assistant'` and the `content` key can be a string (markdown/html supported),
-a `FileDataDict` (to represent a file that is displayed in the chatbot - documented below), or a gradio component.
+The `role` key is either `'user'` or `'`assistant'` and the `content` key can be one of the following:
```
Comment (abidlabs) [inline js/_website/src/lib/templates/gradio/03_components/chatbot.svx:48]:
```suggestion
The `role` key is either `'user'` or `'assistant'` and the `content` key can be one of the following:
```

**Comment by abidlabs** [comment]
(5) The Chatbot page in the docs is pretty good. But it is a bit long. I would recommend moving the part about tools (screenshotted below) to the Examples section below since its more niche. 

image

File: js/_website/src/lib/templates/gradio/03_components/chatbot.svx (hunk 1 above)
Comment (aliabd) [reply js/_website/src/lib/templates/gradio/03_components/chatbot.svx:50]: I use lists where i don't want them to have markers (on navbars and search results etc) so there was a style clash i didn't notice. Fixed now

File: gradio/chat_interface.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/chat_interface.py:95]: 2 is supported. Will reword as suggested

File: gradio/components/chatbot.py [hunk 5]
```diff
@@ -207,6 +208,11 @@ def __init__(
         if type == "messages":
             self.data_model = ChatbotDataMessages
         else:
+            # DeprecationWarning gets filtered out by default
+            warnings.warn(
+                "The 'tuples' format is deprecated and will be removed in version 6 of Gradio. Please use the 'messages' format instead.",
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply gradio/components/chatbot.py:213]: Wouldn't 6 be the next major?

[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/components/chatbot.py:213]: Yeah but just to keep our options open :D idk

File: js/_website/src/lib/templates/gradio/03_components/chatbot.svx (hunk 1 above)
Comment (freddyaboulton) [reply js/_website/src/lib/templates/gradio/03_components/chatbot.svx:58]: Thanks @aliabd !

**Comment by freddyaboulton** [comment]
Thank you for the review @abidlabs and @aliabd for helping with some formatting/docs questions. I think all of the comments have been addressed.

File: gradio/chat_interface.py [hunk 6]
```diff
@@ -183,6 +184,7 @@ def __init__(
                 Markdown(description)
 
             if chatbot:
+                breakpoint()
```
Comment (abidlabs) [inline gradio/chat_interface.py:187]:
```suggestion
```

File: gradio/components/chatbot.py [hunk 7]
```diff
@@ -207,6 +208,11 @@ def __init__(
         if type == "messages":
             self.data_model = ChatbotDataMessages
         else:
+            # DeprecationWarning gets filtered out by default
+            warnings.warn(
+                "The 'tuples' format for chatbot messages is deprecated and will be removed in a future version Gradio. Please set type='messages' instead.",
```
Comment (abidlabs) [inline gradio/components/chatbot.py:213]:
```suggestion
                "The 'tuples' format for chatbot messages is deprecated and will be removed in a future version of Gradio. Please set type='messages' instead.",
```

File: js/_website/src/lib/templates/gradio/03_components/chatbot.svx [hunk 8]
```diff
@@ -38,145 +46,108 @@ demo.launch()
...
-This parameter can take two values, `'tuples'` and `'messages'`.
-
-
-If `type` is `'tuples'`, then the data sent to/from the chatbot will be a list of tuples.
-The first element of each tuple is the user message and the second element is the bot's response.
-Each element can be a string (markdown/html is supported),
-a tuple (in which case the first element is a filepath that will be displayed in the chatbot),
-or a gradio component (see the Examples section for more details).
+This parameter can take two values, `'tuples'` and `'messages'`. 
+The `'tuples'` type is deprecated and will be removed in version 6.0 of Gradio.
```
Comment (abidlabs) [inline js/_website/src/lib/templates/gradio/03_components/chatbot.svx:50]:
```suggestion
The `'tuples'` type is deprecated and will be removed in a future version of Gradio.
```

File: guides/05_chatbots/01_creating-a-chatbot-fast.md [hunk 9]
```diff
@@ -20,14 +20,26 @@ $ pip install --upgrade gradio
...
 - `message`: a `str` representing the user's input.
-- `history`: a `list` of `list` representing the conversations up until that point. Each inner list consists of two `str` representing a pair: `[user input, bot response]`.
+- `history`: If you set type="messages" in gr.ChatInterface, the history will be a list of dictionaries with `role` and `content` keys. Please see the chatbot [docs](/docs/gradio/chatbot) for an in-depth explanation of the chatbot format. 
+
+Here is an example value of the `history`:
+
+```python
+[
+    {"role": "user", "content": "What is the capital of France"},
+    {"role": "assistant", "content": "France"}
```
Comment (abidlabs) [inline guides/05_chatbots/01_creating-a-chatbot-fast.md:33]:
```suggestion
    {"role": "assistant", "content": "Paris"}
```

File: guides/05_chatbots/01_creating-a-chatbot-fast.md [hunk 10]
```diff
@@ -20,14 +20,26 @@ $ pip install --upgrade gradio
 
 ## Defining a chat function
 
-When working with `gr.ChatInterface()`, the first thing you should do is define your chat function. Your chat function should take two arguments: `message` and then `history` (the arguments can be named anything, but must be in this order).
+When working with `gr.ChatInterface()`, the first thing you should do is define your chat function. Your chat function should take two arguments: `message` and `history` (the arguments can be named anything, but must be in this order).
 
 - `message`: a `str` representing the user's input.
-- `history`: a `list` of `list` representing the conversations up until that point. Each inner list consists of two `str` representing a pair: `[user input, bot response]`.
+- `history`: If you set type="messages" in gr.ChatInterface, the history will be a list of dictionaries with `role` and `content` keys. Please see the chatbot [docs](/docs/gradio/chatbot) for an in-depth explanation of the chatbot format.
```
Comment (abidlabs) [inline guides/05_chatbots/01_creating-a-chatbot-fast.md:26]:
```suggestion
- `history`: If you set `type="messages"` in gr.ChatInterface, the history will be a list of dictionaries with `role` and `content` keys. Please see the chatbot [docs](/docs/gradio/chatbot) for an in-depth explanation of the chatbot format. 
```

File: guides/05_chatbots/01_creating-a-chatbot-fast.md [hunk 11]
```diff
@@ -20,14 +20,26 @@ $ pip install --upgrade gradio
...
+    {"role": "user", "content": "What is the capital of France"},
+    {"role": "assistant", "content": "France"}
+]
+```
+
 
 Your function should return a single string response, which is the bot's response to the particular user input `message`. Your function can take into account the `history` of messages, as well as the current message.
 
-Let's take a look at a few examples.
+Tip: It's strongly recommended to set type="messages" in gr.ChatInterface. Setting type="tuples" is deprecated and will be removed in the next version of Gradio.
```
Comment (abidlabs) [inline guides/05_chatbots/01_creating-a-chatbot-fast.md:40]:
```suggestion
Tip: It's strongly recommended to set type="messages" in gr.ChatInterface. Setting type="tuples" is deprecated and will be removed in a future version of Gradio.
```

**Comment by abidlabs** [comment]
Nit: I think there should be more space after the third item in the list and the start of the next paragraph. Also, we can remove the empty lines in the code example to take up less vertical space

image

**Review by abidlabs — APPROVED** [review]
nice @freddyaboulton! left a few nits above, but overall looks good

File: guides/05_chatbots/01_creating-a-chatbot-fast.md (hunk 9 above)
Comment (freddyaboulton) [reply guides/05_chatbots/01_creating-a-chatbot-fast.md:33]: Can you tell I was tired? LOL

**Comment by freddyaboulton** [comment]
Thank you @abidlabs!

**Pull request merged**